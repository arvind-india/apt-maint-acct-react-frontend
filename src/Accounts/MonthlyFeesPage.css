.grid {
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(180px, 1fr));
  grid-auto-rows: minmax(180px, 1fr);
  grid-auto-flow: row;
  grid-gap: 20px;
}
.box {
  background-color: #444;
  color: #fff;
  border-radius: 5px;
  padding: 0;
}
.moYr {
  display: flex;
}
/*
.box:nth-child(even) {
  background-color: #ccc;
  color: #000;
}
*/
/*
.grid li {
  border: 1px solid rgb(0, 95, 107);
  border-radius: 3px;
  background-color: rgba(0, 95, 107, 0.8);
  padding: 0.2em;
  color: #fff;
  display: grid;
  grid-template-rows: 1fr 1fr 1fr;
}

.flat-number {
  grid-row: 1;
  justify-self: start;
  align-self: start;
}
.payment {
  grid-row: 2;
  justify-self: center;
  align-self: center;
}
.paid-date {
  display: flex;
  grid-row: 3;
  justify-self: end;
  align-self: end;
  font-size: 70%;
}
.recorded-at {
    border: none;
    background: none;
    color: #fff;
    font-size: 100%;
}
*/
