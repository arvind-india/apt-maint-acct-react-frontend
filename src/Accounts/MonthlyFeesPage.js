import React from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import { Table } from 'reactstrap'
import { Router } from 'react-router-dom'
import Select from 'react-select';
import {
  MdAdd,
  MdVisibility,
  MdEdit,
  MdDelete
} from 'react-icons/lib/md' // material design icons
import {
  Button,
  Input,
  Label
} from 'reactstrap'

import { history } from '../_helpers'
import { PrivateRoute, FlashMessage } from '../_components'
import { default as detailsPage } from './AccountDetailsPage'
import { DEFAULTS, MONTHS } from '../_constants'
import './MonthlyFeesPage.css'
import {
  accountMonthlyActions as actions,
  userActions,
  durationActions,
  flatActions
} from '../_actions'
import { Fee } from './Fee'

let url = '/accounts'
let module = 'accounts'

export class MonthlyFees extends React.Component {

  constructor(props) {
    super(props)
    this.today = new Date()
    this.state = {
      forMonth: this.today.getMonth(),
      forYear: this.today.getFullYear(),
    }
    this.handleChange = this.handleChange.bind(this)
    //this.getMonthlyAccounts = this.getMonthlyAccounts.bind(this)
    this.handleAccountPayment = this.handleAccountPayment.bind(this)
    this.handleAccountUpdate = this.handleAccountUpdate.bind(this)
    this.handleAccountDelete = this.handleAccountDelete.bind(this)
  }

  componentDidMount() {
    this.getMonthlyAccounts()
    this.props.getAllFlats()
  }

  render() {
    const { flats, alert, authzn, trackHistory, accounts } = this.props
    let hist = trackHistory?history:{}
    return (
      <div>
        <h3>Monthly Maintenance Fees Collection</h3>
        {alert.message && <FlashMessage text={alert.message} color={alert.color} delay={2100}/>}
        <div className="moYr">
          { this.showMonth() }
          { this.showYear()}
        </div>
        { accounts && accounts.loading && <div>loading...</div>}
        { accounts.items && flats.items && authzn && this.showFlatsGrid() }
      </div>
    )
  }

  showMonth() {
    const { forMonth } = this.state
    return <Select
      id="forMonth"
      name="forMonth"
      value={forMonth+1}
      multi={false}
      joinValues={false}
      simpleValue={true}
      placeholder="Select a Month..."
      onChange={this.handleChange}
      valueKey="number"
      labelKey="name"
      options={MONTHS}
    />
  }

  showYear() {
    const { forYear } = this.state
    return <Input
      id="forYear"
      type="number"
      name="forYear"
      value={forYear}
      placeholder="Year here"
      className="inputField"
      min="2013"
      max="2030"
      onChange={this.handleChange}
    />
  }

  handleChange(event) {
    const { name, value } = event.target
    this.setState( { [name]: value }, this.getMonthlyAccounts )
  }
  showFlatsGrid(){
    const { flats, authzn, accounts } = this.props
    console.log('MonthlyFeesPage::showFlatsGrid().........', accounts)
    return <ul className="grid">
      { flats.items.map((flat, index) => {
          let account = accounts.items.find((acct) => acct.flat_number === flat.flat_number)
          return <li
            key={flat.id}
            className="box"
            role="button">
              <Fee
                flatNumber={flat.flat_number}
                account={account}
                authzn={authzn}
                onPayment={() => this.handleAccountPayment(account)}
                onUpdate={(acct) => this.handleAccountUpdate(acct)}
                onDelete={() => this.handleAccountDelete(account.id)}
              />
          </li>
        })
      }
    </ul>
  }
  moYrData() {
    const { forMonth, forYear } = this.state
    return {
      month: forMonth+1,
      year: forYear
    }
  }
  getMonthlyAccounts() {
    //const { forMonth, forYear } = this.state
    this.props.getMonthlyAccountsFor(this.moYrData())
  }
  handleAccountPayment(account) {
    if(window.confirm('Are you sure to add payment?')) {
      this.props.saveChangesAndGetMonthlyList(account, this.moYrData())
    }
  }
  handleAccountUpdate(account) {
    this.props.saveChangesAndGetMonthlyList(account)
  }
  handleAccountDelete(accountId) {
    if(window.confirm('Are you sure to Remove?')) {
      let result = this.props.delete(accountId)
      console.log('result of handleAccountDelete: ', result)
      //this.getMonthlyAccounts()
    }
  }

} // end of class

function mapStateToProps(state) {
  const { accounts, alert, authorizations, users, flats } = state
  const authzn = authorizations[module]
  const trackHistory = true  // added for unit testing; snapshot to be precise
  return {
    accounts,
    alert,
    authzn,
    users,
    flats,
    trackHistory
  }
}

function mapDispatchToProps(dispatch) {
  return {
    getAllFlats: () => {
      dispatch(flatActions.getAll())
    },
    getMonthlyAccountsFor: (data) => {
      dispatch(actions.getMonthlyListFor(data))
    },
    getActive: (key, date) => {
      dispatch(durationActions.getActive(key, date))
    },
/*    saveChanges: (model) => {
      dispatch(actions.saveChanges(model))
    }, */
    saveChangesAndGetMonthlyList: (model, data) => {
      dispatch(actions.saveChangesAndGetMonthlyList(model, data))
    },
    delete: (id) => {
      dispatch(actions.delete(id))
    }
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(MonthlyFees)
