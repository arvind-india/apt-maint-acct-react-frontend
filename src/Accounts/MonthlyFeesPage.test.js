// MonthlyFeesPage.test.js

import React from 'react'
import { shallow } from 'enzyme'

import { MonthlyFees } from './MonthlyFeesPage'

const mockProps = {
  trackHistory: false,  // for unit testing history is NOT required as it has 'keys' that are dynamic
  accounts: {
    loading: false,
    error: false,
    items: [
      {
        id: 1,
        recorded_at: '2018-04-04',
        item: 'maintenance fee',
        flat_number: 'G1',
        name: 'tester',
        for_month: '04',
        for_year: '2018',
        crdr: 'cr',
        amount: '600',
        balance: '',
        category: 'maintenance',
        remarks: 'maintenance for the month april 2018'
      },
      {
        id: 2,
        recorded_at: '2018-04-01',
        item: 'sewage clearance',
        flat_number: 'NA',
        name: 'Sewage Truck Service',
        for_month: '04',
        for_year: '2018',
        crdr: 'dr',
        amount: '2200',
        balance: '',
        category: 'sewage disposal',
        remarks: 'septic tank clearance'      },
      {
        id: 3,
        recorded_at: '2018-03-31',
        item: 'Garbage clearance',
        flat_number: 'NA',
        name: 'Panchayat Garbage disposal Service',
        for_month: '04',
        for_year: '2018',
        crdr: 'dr',
        amount: '200',
        balance: '',
        category: 'maintenance',
        remarks: 'Garbage clearance'
      }
    ]
  },
  users: {
    items: []
  },
  flats: {
    items: []
  },
  alert: {message: "testingAlert"},
  authzn: {
    allowsAdd: true,
    allowsEdit: true,
    allowsDelete: true,
    condition: ''
  },
  getAllFlats: jest.fn(),
  delete: jest.fn(),
  getMonthlyAccountsFor: jest.fn(),
  getActive: jest.fn(),
  saveChanges: jest.fn()
}

const component = shallow(<MonthlyFees {...mockProps} />)

describe('MonthlyFeesPage params test', () => {
  it('should render MonthlyFeesPage without crashing', () => {
    component
  })

  it('should render initial layout', () => {
    expect(component.getElements()).toMatchSnapshot()
  })
})

/*
describe('MonthlyFeesPage rows test', () => {
  let numberOfMockMonthlyFees = mockProps.accounts.items.length
  let tbody = component.find('tbody')
  let rows = tbody.find('tr')
  it('should have '+numberOfMockMonthlyFees+' rows', () => {
    expect(rows.length).toEqual(numberOfMockMonthlyFees)
  })
})
*/
