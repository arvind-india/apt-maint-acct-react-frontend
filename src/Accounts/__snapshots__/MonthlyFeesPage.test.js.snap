// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonthlyFeesPage params test should render initial layout 1`] = `
Array [
  <div>
    <h3>
        Monthly Maintenance Fees Collection
    </h3>
    <FlashMessage
        color={undefined}
        delay={2100}
        text="testingAlert"
    />
    <div
        className="moYr"
    >
        <Select
            arrowRenderer={[Function]}
            autosize={true}
            backspaceRemoves={true}
            backspaceToRemoveMessage="Press backspace to remove {label}"
            clearAllText="Clear all"
            clearRenderer={[Function]}
            clearValueText="Clear value"
            clearable={true}
            closeOnSelect={true}
            deleteRemoves={true}
            delimiter=","
            disabled={false}
            escapeClearsValue={true}
            filterOptions={[Function]}
            id="forMonth"
            ignoreAccents={true}
            ignoreCase={true}
            inputProps={Object {}}
            isLoading={false}
            joinValues={false}
            labelKey="name"
            matchPos="any"
            matchProp="any"
            menuBuffer={0}
            menuRenderer={[Function]}
            multi={false}
            name="forMonth"
            noResultsText="No results found"
            onBlurResetsInput={true}
            onChange={[Function]}
            onCloseResetsInput={true}
            onSelectResetsInput={true}
            openOnClick={true}
            optionComponent={[Function]}
            options={
                Array [
                    Object {
                      "name": "Jan",
                      "number": 1,
                    },
                    Object {
                      "name": "Feb",
                      "number": 2,
                    },
                    Object {
                      "name": "Mar",
                      "number": 3,
                    },
                    Object {
                      "name": "Apr",
                      "number": 4,
                    },
                    Object {
                      "name": "May",
                      "number": 5,
                    },
                    Object {
                      "name": "Jun",
                      "number": 6,
                    },
                    Object {
                      "name": "Jul",
                      "number": 7,
                    },
                    Object {
                      "name": "Aug",
                      "number": 8,
                    },
                    Object {
                      "name": "Sep",
                      "number": 9,
                    },
                    Object {
                      "name": "Oct",
                      "number": 10,
                    },
                    Object {
                      "name": "Nov",
                      "number": 11,
                    },
                    Object {
                      "name": "Dec",
                      "number": 12,
                    },
                  ]
            }
            pageSize={5}
            placeholder="Select a Month..."
            removeSelected={true}
            required={false}
            rtl={false}
            scrollMenuIntoView={true}
            searchable={true}
            simpleValue={true}
            tabSelectsValue={true}
            trimFilter={true}
            value={4}
            valueComponent={[Function]}
            valueKey="number"
        />
        <Input
            className="inputField"
            id="forYear"
            max="2030"
            min="2013"
            name="forYear"
            onChange={[Function]}
            placeholder="Year here"
            tag="p"
            type="number"
            value={2018}
        />
    </div>
    <ul
        className="grid"
    >
        
    </ul>
</div>,
]
`;
